### PR TITLE
[Autodiff] Adds logic to rewrite call-sites using functions specialized by the closure-spec optimization

### PR DESCRIPTION
**Pull request by jkshtj**
This PR adds the final part of the initial autodiff specific closure-specialization optimization implementation. 

The previous parts of the optimization have been implemented in:
1. [Adds part of the Autodiff specific closure-specialization optimization pass](https://github.com/apple/swift/pull/71775)
2. [Adds logic to generate specialized functions in the closure-spec pass](https://github.com/jkshtj/swift/pull/1)

### CODE REVIEW

**Comment by jkshtj** [comment]
@swift-ci please test

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ClosureSpecialization.swift [hunk 1]
```diff
@@ -822,14 +995,30 @@ private extension Builder {
     return (finalClonedReabstractedClosure as! SingleValueInstruction, releasableClonedReabstractedClosures)
   }
 
-  func destroyPartialApplyOnStack(paiOnStack: PartialApplyInst) {
+  func destroyPartialApplyOnStack(paiOnStack: PartialApplyInst) -> Bool {
```
[the remark this answers is not in the document]
Comment (jkshtj) [reply SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ClosureSpecialization.swift:998]: Not needed after your suggested changes.

File: SwiftCompilerSources/Sources/SIL/Function.swift [hunk 2]
```diff
@@ -39,6 +39,8 @@ final public class Function : CustomStringConvertible, HasShortDescription, Hash
 
   public var isAutodiffVJP: Bool { bridged.isAutodiffVJP() }
 
+  public var isOptimizable: Bool { bridged.isOptimizable() }
```
[the remark this answers is not in the document]
Comment (jkshtj) [reply SwiftCompilerSources/Sources/SIL/Function.swift:42]: Not needed after your suggested changes.

File: lib/SILOptimizer/PassManager/PassPipeline.cpp [hunk 3]
```diff
@@ -735,6 +735,17 @@ static void addMidLevelFunctionPipeline(SILPassPipelinePlan &P) {
   P.addLoopUnroll();
 }
 
+static void addAutodiffClosureSpecializationPassPipeline(SILPassPipelinePlan &P) {
+  P.startPipeline("AutodiffClosureSpecialization");
+  P.addDeadStoreElimination();
```
[the remark this answers is not in the document]
Comment (jkshtj) [reply lib/SILOptimizer/PassManager/PassPipeline.cpp:740]: I was modelling the pipeline after the existing closure-specialization pipeline and kept only the basic optimization passes that I knew could affect autodiff.

That said, those other passes are not strictly needed. I've removed them.

**Comment by jkshtj** [comment]
@swift-ci please test

**Comment by jkshtj** [comment]
@swift-ci please test

**Comment by jkshtj** [comment]
@swift-ci please test macos

File: SwiftCompilerSources/Sources/SIL/Function.swift [hunk 4]
```diff
@@ -38,7 +38,7 @@ final public class Function : CustomStringConvertible, HasShortDescription, Hash
   public var isTrapNoReturn: Bool { bridged.isTrapNoReturn() }
 
   public var isAutodiffVJP: Bool { bridged.isAutodiffVJP() }
-
+
```
Comment (eeckstein) [inline SwiftCompilerSources/Sources/SIL/Function.swift:41]: accidental change?

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp [hunk 5]
```diff
@@ -215,7 +215,6 @@ class SILPerformanceInliner {
   bool isAutoDiffLinearMapWithControlFlow(FullApplySite AI);
 
   bool isTupleWithAllocsOrPartialApplies(SILValue retVal);
-
```
Comment (eeckstein) [inline lib/SILOptimizer/Transforms/PerformanceInliner.cpp:218]: accidental change?

**Review by eeckstein — APPROVED** [review]
lgtm!

**Comment by jkshtj** [comment]
@swift-ci please test macos

File: SwiftCompilerSources/Sources/SIL/Function.swift (hunk 4 above)
Comment (jkshtj) [reply SwiftCompilerSources/Sources/SIL/Function.swift]: Looks like it. Fixed now.

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp (hunk 5 above)
Comment (jkshtj) [reply lib/SILOptimizer/Transforms/PerformanceInliner.cpp]: Looks like it. Fixed now.

**Comment by jkshtj** [comment]
@swift-ci please test

**Comment by jkshtj** [comment]
@swift-ci please test

**Comment by jkshtj** [comment]
@swift-ci please test macos

**Comment by jkshtj** [comment]
@swift-ci please test linux

**Comment by jkshtj** [comment]
@swift-ci please test linux

**Comment by jkshtj** [comment]
@swift-ci please test

**Comment by jkshtj** [comment]
@swift-ci please test

**Comment by jkshtj** [comment]
@swift-ci please test macos

**Comment by jkshtj** [comment]
@shahmishal The macos CI builds seem to have been a bit flaky recently. I've had some rare successful builds but mostly failures. The logs seem to be scattered with messages like these:
```
Test Case '-[SKCoreTests.TaskSchedulerTests testTasksWithElevatedPrioritiesGetExecutedFirst]' started.
/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/sourcekit-lsp/Tests/SKCoreTests/TaskSchedulerTests.swift:89: error: -[SKCoreTests.TaskSchedulerTests testTasksWithElevatedPrioritiesGetExecutedFirst] : XCTAssertTrue failed - [Set([high(9)]), Set([])] did not fulfill predicate
Test Case '-[SKCoreTests.TaskSchedulerTests testTasksWithElevatedPrioritiesGetExecutedFirst]' failed (3.794 seconds).
Test Suite 'TaskSchedulerTests' failed at 2024-05-23 10:39:31.836.
	 Executed 1 test, with 1 failure (0 unexpected) in 3.794 (3.794) seconds
Test Suite 'SourceKitLSPPackageTests.xctest' failed at 2024-05-23 10:39:31.836.
	 Executed 1 test, with 1 failure (0 unexpected) in 3.794 (3.794) seconds
Test Suite 'Selected tests' failed at 2024-05-23 10:39:31.836.
	 Executed 1 test, with 1 failure (0 unexpected) in 3.794 (3.795) seconds

warning: --multiroot-data-file is an *unsupported* option which can be removed at any time; do not rely on it
warning: 'swift-format': found 1 file(s) which are unhandled; explicitly declare them as resources or exclude from the target
    /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swift-format/Sources/SwiftFormat/CMakeLists.txt
warning: 'swift-format': found 1 file(s) which are unhandled; explicitly declare them as resources or exclude from the target
    /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swift-format/Sources/SwiftFormat/CMakeLists.txt
warning: 'swift-format': found 1 file(s) which are unhandled; explicitly declare them as resources or exclude from the target
    /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swift-format/Sources/swift-format/CMakeLists.txt
warning: 'swift-format': found 1 file(s) which are unhandled; explicitly declare them as resources or exclude from the target
    /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swift-format/Sources/swift-format/CMakeLists.txt
warning: failed to retrieve search paths with pkg-config; maybe pkg-config is not installed
warning: couldn't find pc file for sqlite3
```

```
[org.swift.sourcekit-lsp:default] error 2024-05-23 10:54:15.7830 +0000
/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/Info.plist doesn't exist in file system
```

Do you know what's going on here?

**Pull request merged**